### PR TITLE
test(dsg): fix flacky e2e tests

### PR DESCRIPTION
**Pull request by overbit**
Close: #7334

## PR Details

This PR consists of 3 changes:

1. This PR introduce a dynamic wait time before starting e2e tests  based on the actual state of the docker-compose stack instead of a static timeout. 
  With this change, the test run time will also be reduce as it won't need to always wait for 30 seconds before running the actual tests. 🚀
2. Update docker-compose to v2

3. ⚠️ During the investigation, I've noticed that another reason for these e2e tests to fail was the following:

 image
 In order to fix this, in the most simple way, this PR adds a restart policy to the docker-compose stack `server` service to avoid the situation were server failed to start, due to race condition with the migration service, and stayed dead.
With the new policy, if `server` starts and fails due to migrations not being completed, it will automatically be restarted by the docker engine.

### CODE REVIEW

File: packages/data-service-generator/tests/e2e/data-service-generator.e2e.spec.ts [hunk 1]
```diff
@@ -193,8 +192,54 @@ describe("Data Service Generator", () => {
             logger.error(err.message, err);
           });
 
+        // Wait for the server to be ready and for database migration before running tests.
+        // TODO replace with a better docker-compose solution that waits for the migraton to be completd before starting the server
+        logger.info("Waiting for db migration to be ...");
```
Comment (mulygottlieb) [inline packages/data-service-generator/tests/e2e/data-service-generator.e2e.spec.ts:197]: "to be ? ... " :) Did you mean "to be completed..." ?

File: packages/data-service-generator/tests/e2e/data-service-generator.e2e.spec.ts [hunk 2]
```diff
@@ -193,8 +192,54 @@ describe("Data Service Generator", () => {
...
+          );
+          if (migrateContainer.state.indexOf("Exited (0)") !== -1) {
+            migrationCompleted = true;
+            logger.info("migration completed!");
+            break;
+          }
+          await sleep(1000);
+        } while (
+          !migrationCompleted ||
+          startTime + SERVER_START_TIMEOUT < Date.now()
```
Comment (mulygottlieb) [inline packages/data-service-generator/tests/e2e/data-service-generator.e2e.spec.ts:218]: Do you want to use the same timeout (SERVER_START_TIMEOUT) for both the migration and the server start?

**Review by mulygottlieb — APPROVED** [review]

File: packages/data-service-generator/tests/e2e/data-service-generator.e2e.spec.ts (hunk 2 above)
Comment (overbit) [reply packages/data-service-generator/tests/e2e/data-service-generator.e2e.spec.ts:218]: We can make them different but I don't seen any reasons for it at the moment

File: packages/data-service-generator/tests/e2e/data-service-generator.e2e.spec.ts (hunk 1 above)
Comment (overbit) [reply packages/data-service-generator/tests/e2e/data-service-generator.e2e.spec.ts:197]: fixed

**Pull request merged**